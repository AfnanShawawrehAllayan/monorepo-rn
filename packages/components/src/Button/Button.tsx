import React from 'react';
import {
  TouchableOpacity,
  Text,
  StyleSheet,
  TouchableOpacityProps,
  ActivityIndicator,
  ViewStyle,
  TextStyle,
} from 'react-native';
import { useTheme } from 'theme';

export interface ButtonProps extends TouchableOpacityProps {
  title: string;
  variant?: 'primary' | 'secondary' | 'outline' | 'text';
  size?: 'small' | 'medium' | 'large';
  loading?: boolean;
  disabled?: boolean;
  style?: ViewStyle;
  textStyle?: TextStyle;
  leftIcon?: React.ReactNode;
  rightIcon?: React.ReactNode;
  testID?: string;
}

export const Button = ({
  title,
  variant = 'primary',
  size = 'medium',
  loading = false,
  disabled = false,
  style,
  textStyle,
  leftIcon,
  rightIcon,
  testID,
  ...rest
}: ButtonProps): React.ReactElement => {
  const { colors, spacing } = useTheme();

  const getBackgroundColor = () => {
    if (disabled) return colors.disabled;

    switch (variant) {
      case 'primary':
        return colors.primary;
      case 'secondary':
        return colors.secondary;
      case 'outline':
      case 'text':
        return 'transparent';
      default:
        return colors.primary;
    }
  };

  const getTextColor = () => {
    if (disabled) return colors.textDisabled;

    switch (variant) {
      case 'primary':
      case 'secondary':
        return colors.textOnPrimary;
      case 'outline':
        return colors.primary;
      case 'text':
        return colors.primary;
      default:
        return colors.textOnPrimary;
    }
  };

  const getBorderColor = () => {
    if (disabled) return colors.disabled;
    return variant === 'outline' ? colors.primary : 'transparent';
  };

  const getPadding = () => {
    switch (size) {
      case 'small':
        return spacing.s;
      case 'medium':
        return spacing.m;
      case 'large':
        return spacing.l;
      default:
        return spacing.m;
    }
  };

  const styles = StyleSheet.create({
    button: {
      alignItems: 'center',
      backgroundColor: getBackgroundColor(),
      borderColor: getBorderColor(),
      borderRadius: 8,
      borderWidth: variant === 'outline' ? 1 : 0,
      flexDirection: 'row',
      justifyContent: 'center',
      paddingHorizontal: getPadding() * 2,
      paddingVertical: getPadding(),
    },
    text: {
      color: getTextColor(),
      fontSize: size === 'small' ? 14 : size === 'medium' ? 16 : 18,
      fontWeight: '600',
    },
  });

  return (
    <TouchableOpacity 
      style={[styles.button, style]} 
      disabled={disabled || loading} 
      testID={testID}
      {...rest}
    >
      {loading ? (
        <ActivityIndicator color={getTextColor()} />
      ) : (
        <>
          {leftIcon}
          <Text style={[styles.text, textStyle]}>{title}</Text>
          {rightIcon}
        </>
      )}
    </TouchableOpacity>
  );
};
