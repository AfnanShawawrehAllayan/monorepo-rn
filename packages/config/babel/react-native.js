module.exports = {
  presets: ['@react-native/babel-preset'],
  plugins: [
    [
      'module-resolver',
      {
        root: ['.'],
        extensions: [
          '.ios.ts',
          '.android.ts',
          '.ts',
          '.ios.tsx',
          '.android.tsx',
          '.tsx',
          '.jsx',
          '.js',
          '.json',
        ],
        alias: {
          '@components': './src/components',
          '@screens': './src/screens',
          '@navigation': './src/navigation',
          '@utils': './src/utils',
          '@hooks': './src/hooks',
          '@assets': './src/assets',
          '@config': './src/config',
        },
      },
    ],
    'react-native-reanimated/plugin',
  ],
  env: {
    production: {
      plugins: ['transform-remove-console'],
    },
  },
};
