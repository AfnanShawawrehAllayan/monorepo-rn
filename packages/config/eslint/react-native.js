module.exports = {
  root: true,
  extends: [
    'eslint:recommended',
    'plugin:@typescript-eslint/recommended',
    'plugin:react/recommended',
    'plugin:react-hooks/recommended',
    'plugin:react-native/all',
    'plugin:import/errors',
    'plugin:import/warnings',
    'plugin:import/typescript',
    'plugin:jest/recommended',
    'prettier',
  ],
  parser: '@typescript-eslint/parser',
  parserOptions: {
    ecmaFeatures: {
      jsx: true,
    },
    ecmaVersion: 2021,
    sourceType: 'module',
  },
  plugins: ['@typescript-eslint', 'react', 'react-native', 'import', 'jest'],
  settings: {
    react: {
      version: 'detect',
    },
    'import/resolver': {
      typescript: {},
      node: {
        extensions: ['.js', '.jsx', '.ts', '.tsx'],
      },
    },
  },
  rules: {
    // React Native specific rules
    'react-native/no-unused-styles': 'error',
    'react-native/split-platform-components': 'error',
    'react-native/no-inline-styles': 'warn',
    'react-native/no-color-literals': 'warn',
    'react-native/no-raw-text': ['warn', { skip: ['Text'] }],
    'react-native/no-single-element-style-arrays': 'error',

    // TypeScript rules
    '@typescript-eslint/explicit-module-boundary-types': 'warn',
    '@typescript-eslint/no-explicit-any': 'warn',
    '@typescript-eslint/no-unused-vars': ['error', { argsIgnorePattern: '^_' }],
    '@typescript-eslint/consistent-type-imports': ['error', { prefer: 'type-imports' }],

    // React rules
    'react/prop-types': 'off',
    'react/react-in-jsx-scope': 'off',
    'react/display-name': 'off',
    'react/jsx-curly-brace-presence': ['error', { props: 'never', children: 'never' }],

    // Import rules
    'import/order': [
      'error',
      {
        groups: ['builtin', 'external', 'internal', 'parent', 'sibling', 'index'],
        'newlines-between': 'always',
        alphabetize: {
          order: 'asc',
          caseInsensitive: true,
        },
      },
    ],
    'import/no-duplicates': 'error',
    'import/no-unresolved': 'error',
    'import/no-cycle': 'error',

    // Jest rules
    'jest/no-disabled-tests': 'warn',
    'jest/no-focused-tests': 'error',
    'jest/no-identical-title': 'error',
    'jest/valid-expect': 'error',
  },
  env: {
    'react-native/react-native': true,
    jest: true,
    'jest/globals': true,
  },
  overrides: [
    {
      files: ['*.test.ts', '*.test.tsx', '*.spec.ts', '*.spec.tsx'],
      rules: {
        '@typescript-eslint/no-explicit-any': 'off',
        'react-native/no-raw-text': 'off',
      },
    },
    {
      files: ['*.js', '*.jsx'],
      rules: {
        '@typescript-eslint/explicit-module-boundary-types': 'off',
      },
    },
  ],
};
