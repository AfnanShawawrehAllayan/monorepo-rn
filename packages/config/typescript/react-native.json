{
  "$schema": "https://json.schemastore.org/tsconfig",
  "display": "React Native",
  "compilerOptions": {
    "target": "esnext",
    "module": "esnext",
    "lib": ["esnext"],
    "allowJs": true,
    "jsx": "react-native",
    "noEmit": true,
    "isolatedModules": true,
    "strict": true,
    "moduleResolution": "node",
    "allowSyntheticDefaultImports": true,
    "esModuleInterop": true,
    "skipLibCheck": true,
    "resolveJsonModule": true,
    "baseUrl": ".",
    "paths": {
      "*": ["*", "*.ios", "*.android"]
    },
    "forceConsistentCasingInFileNames": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": true,
    "noUncheckedIndexedAccess": true,
    "noImplicitOverride": true,
    "noPropertyAccessFromIndexSignature": true,
    "exactOptionalPropertyTypes": true,
    "types": ["jest", "react-native"]
  },
  "exclude": ["node_modules", "babel.config.js", "metro.config.js", "jest.config.js"]
}
