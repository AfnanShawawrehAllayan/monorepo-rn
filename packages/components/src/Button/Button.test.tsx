import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { Button } from './Button';
import { ActivityIndicator, Text } from 'react-native';

// Mock the theme hook
jest.mock('theme', () => ({
  useTheme: () => ({
    colors: {
      primary: '#007AFF',
      secondary: '#5856D6',
      disabled: '#A9A9A9',
      textOnPrimary: '#FFFFFF',
      textDisabled: '#808080',
    },
    spacing: {
      s: 8,
      m: 16,
      l: 24,
    },
  }),
}));

describe('Button Component', () => {
  it('renders correctly with default props', () => {
    const { getByText, toJSON } = render(<Button title="Press Me" />);
    expect(getByText('Press Me')).toBeTruthy();
    expect(toJSON()).toMatchSnapshot('Button component - default props');
  });

  it('calls onPress when pressed', () => {
    const onPressMock = jest.fn();
    const { getByText, toJSON } = render(
      <Button title="Press Me" onPress={onPressMock} />
    );

    fireEvent.press(getByText('Press Me'));
    expect(onPressMock).toHaveBeenCalledTimes(1);
    expect(toJSON()).toMatchSnapshot('Button component - with onPress handler');
  });

  it('renders correctly when disabled', () => {
    const { getByTestId, toJSON } = render(
      <Button title="Press Me" disabled testID="button" />
    );
    
    const button = getByTestId('button');
    expect(button.props.disabled).toBe(true);
    expect(toJSON()).toMatchSnapshot('Button component - disabled state');
  });

  it('applies custom style', () => {
    const customStyle = { backgroundColor: 'red' };
    const { getByTestId, toJSON } = render(
      <Button title="Press Me" style={customStyle} testID="button" />
    );
    
    const button = getByTestId('button');
    const buttonStyles = button.props.style;
    expect(buttonStyles).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          backgroundColor: 'red',
        }),
      ])
    );
    expect(toJSON()).toMatchSnapshot('Button component - with custom style');
  });



  it('renders with different variants', () => {
    const variants = ['primary', 'secondary', 'outline', 'text'] as const;
    
    variants.forEach(variant => {
      const { getByTestId, toJSON } = render(
        <Button title="Press Me" variant={variant} testID={`button-${variant}`} />
      );
      
      const button = getByTestId(`button-${variant}`);
      expect(button).toBeTruthy();
      expect(toJSON()).toMatchSnapshot(`Button component - ${variant} variant`);
    });
  });

  it('renders with different sizes', () => {
    const sizes = ['small', 'medium', 'large'] as const;
    
    sizes.forEach(size => {
      const { getByTestId, toJSON } = render(
        <Button title="Press Me" size={size} testID={`button-${size}`} />
      );
      
      const button = getByTestId(`button-${size}`);
      expect(button).toBeTruthy();
      expect(toJSON()).toMatchSnapshot(`Button component - ${size} size`);
    });
  });

  it('renders with left and right icons', () => {
    const TestIcon = () => <Text>icon</Text>;
    
    const { getAllByText, toJSON } = render(
      <Button 
        title="Press Me" 
        leftIcon={<TestIcon />}
        rightIcon={<TestIcon />}
      />
    );
    
    const icons = getAllByText('icon');
    expect(icons).toHaveLength(2);
    expect(toJSON()).toMatchSnapshot('Button component - with icons');
  });

  it('applies custom text style', () => {
    const customTextStyle = { fontSize: 20, color: 'blue' };
    const { getByText, toJSON } = render(
      <Button title="Press Me" textStyle={customTextStyle} />
    );
    
    const buttonText = getByText('Press Me');
    expect(buttonText.props.style).toEqual(
      expect.arrayContaining([
        expect.objectContaining(customTextStyle),
      ])
    );
    expect(toJSON()).toMatchSnapshot('Button component - with custom text style');
  });
});
