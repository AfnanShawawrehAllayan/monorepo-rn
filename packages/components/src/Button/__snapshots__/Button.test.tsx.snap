// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component applies custom style: Button component - with custom style 1`] = `
<TouchableOpacity
  disabled={false}
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#007AFF",
        "borderColor": "transparent",
        "borderRadius": 8,
        "borderWidth": 0,
        "flexDirection": "row",
        "justifyContent": "center",
        "paddingHorizontal": 32,
        "paddingVertical": 16,
      },
      {
        "backgroundColor": "red",
      },
    ]
  }
  testID="button"
>
  <Text
    style={
      [
        {
          "color": "#FFFFFF",
          "fontSize": 16,
          "fontWeight": "600",
        },
        undefined,
      ]
    }
  >
    Press Me
  </Text>
</TouchableOpacity>
`;

exports[`Button Component applies custom text style: Button component - with custom text style 1`] = `
<TouchableOpacity
  disabled={false}
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#007AFF",
        "borderColor": "transparent",
        "borderRadius": 8,
        "borderWidth": 0,
        "flexDirection": "row",
        "justifyContent": "center",
        "paddingHorizontal": 32,
        "paddingVertical": 16,
      },
      undefined,
    ]
  }
>
  <Text
    style={
      [
        {
          "color": "#FFFFFF",
          "fontSize": 16,
          "fontWeight": "600",
        },
        {
          "color": "blue",
          "fontSize": 20,
        },
      ]
    }
  >
    Press Me
  </Text>
</TouchableOpacity>
`;

exports[`Button Component calls onPress when pressed: Button component - with onPress handler 1`] = `
<TouchableOpacity
  disabled={false}
  onPress={
    [MockFunction] {
      "calls": [
        [],
      ],
      "results": [
        {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#007AFF",
        "borderColor": "transparent",
        "borderRadius": 8,
        "borderWidth": 0,
        "flexDirection": "row",
        "justifyContent": "center",
        "paddingHorizontal": 32,
        "paddingVertical": 16,
      },
      undefined,
    ]
  }
>
  <Text
    style={
      [
        {
          "color": "#FFFFFF",
          "fontSize": 16,
          "fontWeight": "600",
        },
        undefined,
      ]
    }
  >
    Press Me
  </Text>
</TouchableOpacity>
`;

exports[`Button Component renders correctly when disabled: Button component - disabled state 1`] = `
<TouchableOpacity
  disabled={true}
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#A9A9A9",
        "borderColor": "#A9A9A9",
        "borderRadius": 8,
        "borderWidth": 0,
        "flexDirection": "row",
        "justifyContent": "center",
        "paddingHorizontal": 32,
        "paddingVertical": 16,
      },
      undefined,
    ]
  }
  testID="button"
>
  <Text
    style={
      [
        {
          "color": "#808080",
          "fontSize": 16,
          "fontWeight": "600",
        },
        undefined,
      ]
    }
  >
    Press Me
  </Text>
</TouchableOpacity>
`;

exports[`Button Component renders correctly with default props: Button component - default props 1`] = `
<TouchableOpacity
  disabled={false}
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#007AFF",
        "borderColor": "transparent",
        "borderRadius": 8,
        "borderWidth": 0,
        "flexDirection": "row",
        "justifyContent": "center",
        "paddingHorizontal": 32,
        "paddingVertical": 16,
      },
      undefined,
    ]
  }
>
  <Text
    style={
      [
        {
          "color": "#FFFFFF",
          "fontSize": 16,
          "fontWeight": "600",
        },
        undefined,
      ]
    }
  >
    Press Me
  </Text>
</TouchableOpacity>
`;

exports[`Button Component renders with different sizes: Button component - large size 1`] = `
<TouchableOpacity
  disabled={false}
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#007AFF",
        "borderColor": "transparent",
        "borderRadius": 8,
        "borderWidth": 0,
        "flexDirection": "row",
        "justifyContent": "center",
        "paddingHorizontal": 48,
        "paddingVertical": 24,
      },
      undefined,
    ]
  }
  testID="button-large"
>
  <Text
    style={
      [
        {
          "color": "#FFFFFF",
          "fontSize": 18,
          "fontWeight": "600",
        },
        undefined,
      ]
    }
  >
    Press Me
  </Text>
</TouchableOpacity>
`;

exports[`Button Component renders with different sizes: Button component - medium size 1`] = `
<TouchableOpacity
  disabled={false}
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#007AFF",
        "borderColor": "transparent",
        "borderRadius": 8,
        "borderWidth": 0,
        "flexDirection": "row",
        "justifyContent": "center",
        "paddingHorizontal": 32,
        "paddingVertical": 16,
      },
      undefined,
    ]
  }
  testID="button-medium"
>
  <Text
    style={
      [
        {
          "color": "#FFFFFF",
          "fontSize": 16,
          "fontWeight": "600",
        },
        undefined,
      ]
    }
  >
    Press Me
  </Text>
</TouchableOpacity>
`;

exports[`Button Component renders with different sizes: Button component - small size 1`] = `
<TouchableOpacity
  disabled={false}
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#007AFF",
        "borderColor": "transparent",
        "borderRadius": 8,
        "borderWidth": 0,
        "flexDirection": "row",
        "justifyContent": "center",
        "paddingHorizontal": 16,
        "paddingVertical": 8,
      },
      undefined,
    ]
  }
  testID="button-small"
>
  <Text
    style={
      [
        {
          "color": "#FFFFFF",
          "fontSize": 14,
          "fontWeight": "600",
        },
        undefined,
      ]
    }
  >
    Press Me
  </Text>
</TouchableOpacity>
`;

exports[`Button Component renders with different variants: Button component - outline variant 1`] = `
<TouchableOpacity
  disabled={false}
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "borderColor": "#007AFF",
        "borderRadius": 8,
        "borderWidth": 1,
        "flexDirection": "row",
        "justifyContent": "center",
        "paddingHorizontal": 32,
        "paddingVertical": 16,
      },
      undefined,
    ]
  }
  testID="button-outline"
>
  <Text
    style={
      [
        {
          "color": "#007AFF",
          "fontSize": 16,
          "fontWeight": "600",
        },
        undefined,
      ]
    }
  >
    Press Me
  </Text>
</TouchableOpacity>
`;

exports[`Button Component renders with different variants: Button component - primary variant 1`] = `
<TouchableOpacity
  disabled={false}
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#007AFF",
        "borderColor": "transparent",
        "borderRadius": 8,
        "borderWidth": 0,
        "flexDirection": "row",
        "justifyContent": "center",
        "paddingHorizontal": 32,
        "paddingVertical": 16,
      },
      undefined,
    ]
  }
  testID="button-primary"
>
  <Text
    style={
      [
        {
          "color": "#FFFFFF",
          "fontSize": 16,
          "fontWeight": "600",
        },
        undefined,
      ]
    }
  >
    Press Me
  </Text>
</TouchableOpacity>
`;

exports[`Button Component renders with different variants: Button component - secondary variant 1`] = `
<TouchableOpacity
  disabled={false}
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#5856D6",
        "borderColor": "transparent",
        "borderRadius": 8,
        "borderWidth": 0,
        "flexDirection": "row",
        "justifyContent": "center",
        "paddingHorizontal": 32,
        "paddingVertical": 16,
      },
      undefined,
    ]
  }
  testID="button-secondary"
>
  <Text
    style={
      [
        {
          "color": "#FFFFFF",
          "fontSize": 16,
          "fontWeight": "600",
        },
        undefined,
      ]
    }
  >
    Press Me
  </Text>
</TouchableOpacity>
`;

exports[`Button Component renders with different variants: Button component - text variant 1`] = `
<TouchableOpacity
  disabled={false}
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "borderColor": "transparent",
        "borderRadius": 8,
        "borderWidth": 0,
        "flexDirection": "row",
        "justifyContent": "center",
        "paddingHorizontal": 32,
        "paddingVertical": 16,
      },
      undefined,
    ]
  }
  testID="button-text"
>
  <Text
    style={
      [
        {
          "color": "#007AFF",
          "fontSize": 16,
          "fontWeight": "600",
        },
        undefined,
      ]
    }
  >
    Press Me
  </Text>
</TouchableOpacity>
`;

exports[`Button Component renders with left and right icons: Button component - with icons 1`] = `
<TouchableOpacity
  disabled={false}
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#007AFF",
        "borderColor": "transparent",
        "borderRadius": 8,
        "borderWidth": 0,
        "flexDirection": "row",
        "justifyContent": "center",
        "paddingHorizontal": 32,
        "paddingVertical": 16,
      },
      undefined,
    ]
  }
>
  <Text>
    icon
  </Text>
  <Text
    style={
      [
        {
          "color": "#FFFFFF",
          "fontSize": 16,
          "fontWeight": "600",
        },
        undefined,
      ]
    }
  >
    Press Me
  </Text>
  <Text>
    icon
  </Text>
</TouchableOpacity>
`;
